File: backend/internal/handlers/catalog.go
package handlers

import (
	"net/http"
	"strconv"

	mediakit "mediakit"
	"streamly/internal/services"

	"github.com/gin-gonic/gin"
)

type CatalogHandler struct {

	media *services.MediaService

}

func NewCatalogHandler(media *services.MediaService) *CatalogHandler {

	return &CatalogHandler{media: media}

}

func (h *CatalogHandler) Search(c *gin.Context) {

	query := c.Query("q")

	if query == "" {

		writeError(c, http.StatusBadRequest, "query required")
		return

	}

	results, err := h.media.Search(query)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, jsonSlice(results))

}

func (h *CatalogHandler) MovieTrending(c *gin.Context) {

	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "12"))

	items, err := h.media.TrendingHits(mediakit.MediaMovie, limit)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, jsonSlice(items))

}

func (h *CatalogHandler) ShowTrending(c *gin.Context) {

	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "12"))

	items, err := h.media.TrendingHits(mediakit.MediaShow, limit)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, jsonSlice(items))

}

func (h *CatalogHandler) MovieCategories(c *gin.Context) {

	cats, err := h.media.Categories(mediakit.MediaMovie)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, jsonSlice(cats))

}

func (h *CatalogHandler) ShowCategories(c *gin.Context) {

	cats, err := h.media.Categories(mediakit.MediaShow)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, jsonSlice(cats))

}

func (h *CatalogHandler) MovieCategoryTitles(c *gin.Context) {

	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "24"))

	titles, err := h.media.CategoryTitles(mediakit.MediaMovie, c.Param("id"), page, limit)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, jsonSlice(titles))

}

func (h *CatalogHandler) ShowCategoryTitles(c *gin.Context) {

	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "24"))

	titles, err := h.media.CategoryTitles(mediakit.MediaShow, c.Param("id"), page, limit)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, jsonSlice(titles))

}

func (h *CatalogHandler) MovieDetails(c *gin.Context) {

	id, err := strconv.Atoi(c.Param("id"))

	if err != nil {

		writeError(c, http.StatusBadRequest, "invalid id")
		return

	}

	details, err := h.media.MovieDetails(id)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, details)

}

func (h *CatalogHandler) ShowDetails(c *gin.Context) {

	id, err := strconv.Atoi(c.Param("id"))

	if err != nil {

		writeError(c, http.StatusBadRequest, "invalid id")
		return

	}

	details, err := h.media.ShowDetails(id)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, details)

}

func (h *CatalogHandler) ShowSeasons(c *gin.Context) {

	id, err := strconv.Atoi(c.Param("id"))

	if err != nil {

		writeError(c, http.StatusBadRequest, "invalid id")
		return

	}

	seasons, err := h.media.ShowSeasons(id)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, jsonSlice(seasons))

}

func (h *CatalogHandler) SeasonEpisodes(c *gin.Context) {

	showID, err := strconv.Atoi(c.Param("id"))

	if err != nil {

		writeError(c, http.StatusBadRequest, "invalid show id")
		return

	}

	season, err := strconv.Atoi(c.Param("season"))

	if err != nil {

		writeError(c, http.StatusBadRequest, "invalid season")
		return

	}

	episodes, err := h.media.SeasonEpisodes(showID, season)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, jsonSlice(episodes))

}

func (h *CatalogHandler) EpisodeDetails(c *gin.Context) {

	showID, err := strconv.Atoi(c.Param("id"))

	if err != nil {

		writeError(c, http.StatusBadRequest, "invalid show id")
		return

	}

	season, err := strconv.Atoi(c.Param("season"))

	if err != nil {

		writeError(c, http.StatusBadRequest, "invalid season")
		return

	}

	episode, err := strconv.Atoi(c.Param("episode"))

	if err != nil {

		writeError(c, http.StatusBadRequest, "invalid episode")
		return

	}

	details, err := h.media.EpisodeDetails(showID, season, episode)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, details)

}

func (h *CatalogHandler) LiveChannels(c *gin.Context) {

	channels, err := h.media.LiveChannels()

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, jsonSlice(channels))

}

func (h *CatalogHandler) LivePopular(c *gin.Context) {

	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "24"))

	channels, err := h.media.LivePopular(limit)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, jsonSlice(channels))

}

func (h *CatalogHandler) LiveSearch(c *gin.Context) {

	query := c.Query("q")

	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "48"))

	channels, err := h.media.LiveSearch(query, limit)

	if err != nil {

		handleServiceError(c, err)
		return

	}

	c.JSON(http.StatusOK, jsonSlice(channels))

}

File: Frontend/src/Net/Catalog.ts
import type { Category, Episode, HomeFeed, ResolveResult, SearchHit, Season, TitleDetails } from "@/Types";

import { request } from "./Request";

export const catalogAPI = {

  search(q: string) {

    return request<SearchHit[]>(`/api/search?q=${encodeURIComponent(q)}`);

  },

  homeFeed(kind: "movie" | "show") {

    return request<HomeFeed>(kind === "movie" ? "/api/feed/movies" : "/api/feed/shows");

  },

  resolveTitle(item: { tmdbId?: number; kind: "movie" | "show"; title?: string; year?: number }) {

    const params = new URLSearchParams({

      kind: item.kind,
      tmdbId: String(item.tmdbId ?? 0),

    });

    if (item.title) params.set("title", item.title);

    if (item.year) params.set("year", String(item.year));

    return request<ResolveResult>(`/api/resolve?${params}`);

  },

  movieTrending(limit = 12) {

    return request<SearchHit[]>(`/api/movies/trending?limit=${limit}`);

  },

  showTrending(limit = 12) {

    return request<SearchHit[]>(`/api/shows/trending?limit=${limit}`);

  },

  movieCategories() {

    return request<Category[]>("/api/movies/categories");

  },

  showCategories() {

    return request<Category[]>("/api/shows/categories");

  },

  movieCategoryTitles(id: string, page = 1) {

    return request<SearchHit[]>(`/api/movies/categories/${id}?page=${page}&limit=24`);

  },

  showCategoryTitles(id: string, page = 1) {

    return request<SearchHit[]>(`/api/shows/categories/${id}?page=${page}&limit=24`);

  },

  movieDetails(id: number) {

    return request<TitleDetails>(`/api/movies/${id}`);

  },

  showDetails(id: number) {

    return request<TitleDetails>(`/api/shows/${id}`);

  },

  showSeasons(id: number) {

    return request<Season[]>(`/api/shows/${id}/seasons`);

  },

  seasonEpisodes(showId: number, season: number) {

    return request<Episode[]>(`/api/shows/${showId}/seasons/${season}/episodes`);

  },

  episodeDetails(showId: number, season: number, episode: number) {

    return request<Episode>(`/api/shows/${showId}/seasons/${season}/episodes/${episode}`);

  },

};
